src/top.rs

```diff
@@ -334,6 +345,21 @@ impl Link {
             }
         };
 
+        match message {
+            // Duplicate UDP messages
+            Protocol::Udp(ref datagram)
+                if self.rand_duplication(global_config.message_duplication(), rand) =>
+            {
+                self.sent.push_back(Sent {
```

Comment:
I'd like to get @camshaft's take here on what would be useful.

Reply:
In s2n-quic, [we duplicate before we apply any delay or corruption](https://github.com/aws/s2n-quic/blob/f7dd7844486141acdf4ecbf68feb752c042d191f/quic/s2n-quic-platform/src/io/testing/model.rs#L323-L334) in order to have each packet be treated separately. This enables the simulation to interleave duplicate packets with others rather than having a burst of duplicate packets since [each packet gets its own delivery time](https://github.com/aws/s2n-quic/blob/f7dd7844486141acdf4ecbf68feb752c042d191f/quic/s2n-quic-platform/src/io/testing/model.rs#L309). We also loop up to 5 times instead of a single duplication, may be useful depending on the application under test.

Reply:
Do you think it makes sense to make this configurable with a single float? For example if you specify duplication 5.0 it duplicates five times and 5.5 would be between five and six.

Reply:
it depends on what the intent is... with the s2n-quic implementation each iteration is exponentially less likely to happen since the loop bails on the first retransmission fails:

```rust
            let mut count = 0;
            while count < 5 && gen_rate(retransmit_rate) {
                transmission_count += transmit(Cow::Borrowed(&packet));
                count += 1;
            }

            transmission_count += transmit(Cow::Owned(packet));
```

If you have a single float, it wouldn't be the same thing. Not saying it needs to be the same. But I'm not sure how useful it is to always duplicate packets in fixed bursts like that. IMO the exponential backoff is closer to how real networks behave.